# Cleaning and analyzing eBay Car Sales Data

<img src="./images/car_sales.jpg"/>

This notebook has the objetive to cleaning and analyzing the **eBay Car Sales Dataset**. 

## Group Members:

- Arthur França Pessoa da Cunha Lima
- Thiago Maia Souto

## 0. The Dataset

The dataset used consists in a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website and was originally scraped and uploaded to Kaggle by user orgesleka. However, the original dataset isn't avaliable on Kaggle anymore, but can be found [here](https://data.world/data-society/used-cars-data]). 

Moreover, the original dataset was modified by *DataQuest Team* for educational purposes. The modifications consists in sampling 50,000 data points from the full dataset and was dirtied a bit to more closely resemble what be expected from a scraped dataset. 

The data dictionary from the dataset is as follows : 

- ``dateCrawled ``- When this ad was first crawled. All field-values are taken from this date.
- ``name`` - Name of the car.
- ``seller`` - Whether the seller is private or a dealer.
- ``offerType`` - The type of listing
- ``price`` - The price on the ad to sell the car.
- ``abtest`` - Whether the listing is included in an A/B test.
- ``vehicleType`` - The vehicle Type.
- ``yearOfRegistration`` - The year in which the car was first registered.
- ``gearbox`` - The transmission type.
- ``powerPS`` - The power of the car in PS.
- ``model`` - The car model name.
- ``kilometer`` - How many kilometers the car has driven.
- ``monthOfRegistration`` - The month in which the car was first registered.
- ``fuelType`` - What type of fuel the car uses.
- ``brand`` - The brand of the car.
- ``notRepairedDamage`` - If the car has a damage which is not yet repaired.
- ``dateCreated`` - The date on which the eBay listing was created.
- ``nrOfPictures`` - The number of pictures in the ad.
- ``postalCode`` - The postal code for the location of the vehicle.
- ``lastSeenOnline`` - When the crawler saw this ad last online.


## 1. Introduction

Importing the libraries

In [92]:
import numpy as np
import pandas as pd

Reading the dataset in the variable ``autos``

In [93]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

Obtaining more information about the dataset

In [94]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [95]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [96]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

Observing the information above it's possible to note:
- The dataset contains 20 columns and most of then are string ;
- The columns **vehicleType**, **gearbox**, **model**, **fuelType**, **notRepairedDamage** has missing data;
- The columns name are in camelcase pattern.
- The columns **dateCrawled**, **dateCreated** and **lastSeen** are of type *date* but were recognized as *string* 

## 2. Cleaning Column Names

Once information about the data has been obtained, it's important to change the columns names to make them easier to work with.

In [97]:
autos_columns = autos.columns
print(autos_columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [98]:
autos.rename(columns = {"yearOfRegistration": "registration_year",
                       "monthOfRegistration": "registration_month",
                       "notRepairedDamage": "unrepaired_damage",
                       "dateCreated": "ad_created",
                       "dateCrawled": "date_crawled",
                       "offerType" : "offer_type",
                       "vehicleType": "vehicle_type",
                       "powerPS" : "power_ps",
                       "fuelType" : "fuel_type",
                       "nrOfPictures": "nr_of_pictures",
                        "postalCode" : "postal_code",
                        "lastSeen" : "last_seen"
                       }, inplace = True)

In [99]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Some modification was maded in the columns:
- Changing the columns name from camelcase to snakecase;
- Changing some columns name for better describe the information in that columns:
    - ``yearOfRegistration`` to ``registration_year``;
    - ``monthOfRegistration`` to ``registration_month``;
    - ``notRepairedDamage`` to ``unrepaired_damage``;
    - ``dateCreated`` to ``ad_created``.

## 3. Initial Exploration and Cleaning

The next step is better exploring the data searching for things that need to be fixed.

In [100]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Observing the informantion above is possible to noting that:
- The column **nr_of_pictures** doesnt have any valueable information because all values are 0.
- The column **seller** has seemingly irrelevant information, because has two possible  values and one of them just seems once.
- The column **offer_type** has seemingly irrelevant information, because has two possible  values and one of them just seems once.
- The column **registration_month** need more investigation because the minimum month is 0, but that isn't a valid value for month.
- The column **registration_year** need more investigation because the minimum value is 1000 and the maximum value is 9999 and they aren't valid values for year, since the cars are invented in 1886 and 9999 is so much in the future.
- The column **power_ps** need more investigation because the maximum value is 17700ps and the most powerful car currently is of the order of 2000ps. 
- The column **price** need be convert to numeric dtype.
- The column **odometer** need be convert to numeric dtype.


In [101]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

It can be seen from above that column ``nr_of_pictures`` has only one possible value thus it can be dropped from the dataframe since don't have relevant information.

In [102]:
autos.drop("nr_of_pictures", axis = 1, inplace = True)

In [103]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

Using the same argument as the column ``nr_of_pictures``, the column ``seller`` can be dropped too because almost all values are the same.

In [104]:
autos.drop("seller", axis = 1, inplace = True)

In [105]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

For the same reason as the columns ``nr_of_pictures`` and ``seller`` were dropped, the column ``offer_type`` can be dropped for not contain relevant information 

In [106]:
autos.drop("offer_type", axis = 1, inplace = True)

Moreover, now will be analyzed the columns that need more investigation

In [107]:
autos["registration_month"].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  0,  9, 11,  5,  2,  1], dtype=int64)

From the output above is possible to see that column ``registration_month`` has an invalid value of 0, thus some strategy must be applied to decide what to do with rows with value 0 in this column later.

In [108]:
autos["registration_year"].unique()

array([2004, 1997, 2009, 2007, 2003, 2006, 1995, 1998, 2000, 2017, 2010,
       1999, 1982, 1990, 2015, 2014, 1996, 1992, 2005, 2002, 2012, 2011,
       2008, 1985, 2016, 1994, 1986, 2001, 2018, 2013, 1972, 1993, 1988,
       1989, 1967, 1973, 1956, 1976, 4500, 1987, 1991, 1983, 1960, 1969,
       1950, 1978, 1980, 1984, 1963, 1977, 1961, 1968, 1934, 1965, 1971,
       1966, 1979, 1981, 1970, 1974, 1910, 1975, 5000, 4100, 2019, 1959,
       9996, 9999, 6200, 1964, 1958, 1800, 1948, 1931, 1943, 9000, 1941,
       1962, 1927, 1937, 1929, 1000, 1957, 1952, 1111, 1955, 1939, 8888,
       1954, 1938, 2800, 5911, 1500, 1953, 1951, 4800, 1001], dtype=int64)

It can be seen from above that has invalid values of years in the column ``registration_year`` and thus some strategy must be applied to deal with it later. 

The next step is convert to numeric type all data that was stored as text. Starting with the analysis of patterns in data values from column ``price``

In [109]:
autos["price"]

0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: object

Using the string pattern discovered now is possible convert the dtype of column ``price`` to integer.

In [110]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
autos["price"].head()

C:\Users\THIAGO~1\AppData\Local\Temp/ipykernel_16832/490306896.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)


0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32

Analyzing the patterns in data values from column ``odometer``.

In [111]:
autos["odometer"]

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object

Using the string pattern discovered now is possible convert the dtype of column ``odometer`` to integer.

In [112]:
autos["odometer"] = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
autos["odometer"].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int32

For describing better the information in the column ``odometer`` it was renamed to ``odometer_km`` to include the information that values are in kilometers.

In [113]:
autos.rename(columns = {"odometer": "odometer_km"}, inplace = "True")

## 4. Exploring the Odometer and Price Columns

Once the conversion of dtype of columns ``price`` and ``odometer_km`` from text to integer has been made, it is necessary to perform an exploration in that columns.

###  4.1 Analyzing the column **price**

In first place is necessary obtain more information about the amount o unique values and the statistical distribution of these values in the column

In [114]:
print(autos["price"].unique().shape)
print(autos["price"].describe())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


Observing the information above is possible to noting that minimum price is \\$0 and the maximum is \\$1000000000, obviously  something is wrong with these values. The next step is obtain more information about the distribution, for this purpose will be analyzed the values and their frequencies of appearance in the column will be analyzed starting with the lowest values 

In [115]:
autos["price"].value_counts().head(10)

0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64

As can be seen above, the value zero is the most frequently price among the possible prices. Now it is necessary to look the other extreme, the highest values

In [116]:
autos["price"].value_counts().sort_index(ascending = False).head(10)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64

Observing the above values it is possible to notice that the highest values for the price of a car are really very large, the highest being \\$99999999. Once this information about the data has been obtained, it's important to remove these outliers. For this reason a price in the range of \\$500 to \\$1300000 has been chosen.

In [117]:
autos = autos[autos["price"].between(500,1300000)]
autos["price"].describe()

count    4.510200e+04
mean     6.442671e+03
std      1.489943e+04
min      5.000000e+02
25%      1.500000e+03
50%      3.500000e+03
75%      7.900000e+03
max      1.300000e+06
Name: price, dtype: float64

### 4.2 Analyzing the column **odometer_km**

Obtaining more information about the amount o unique values and the statistical distribution of these values in the column

In [118]:
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())

(13,)
count     45102.000000
mean     125293.002528
std       39622.647172
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


Since the statistical values from the column seems reasonable, it is necessary obtain more information about the frequency distribution of the values.

In [119]:
autos["odometer_km"].value_counts().sort_index().head(13)

5000        616
10000       237
20000       727
30000       765
40000       808
50000       997
60000      1131
70000      1189
80000      1385
90000      1676
100000     2031
125000     4839
150000    28701
Name: odometer_km, dtype: int64

From the output above, it can be noted that values of the column ``odometer_km`` are in the range of 5000km to 150000km, which are reasonable values. Thus don't be necessary to remove any outlier.

## 5. Exploring the date columns

### 5.1. Understanding how the values in the columns 'date_crawled', 'ad_created', 'last_seen' are formatted

Calculating the distribution of values in the column ``date_crawled`` as percentages

In [120]:
autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.025564
2016-03-06    0.014124
2016-03-07    0.036185
2016-03-08    0.033169
2016-03-09    0.032903
2016-03-10    0.032704
2016-03-11    0.033014
2016-03-12    0.037315
2016-03-13    0.015520
2016-03-14    0.036296
2016-03-15    0.034012
2016-03-16    0.029356
2016-03-17    0.031174
2016-03-18    0.012882
2016-03-19    0.034743
2016-03-20    0.038069
2016-03-21    0.037759
2016-03-22    0.033036
2016-03-23    0.032393
2016-03-24    0.028979
2016-03-25    0.031085
2016-03-26    0.032637
2016-03-27    0.031174
2016-03-28    0.034832
2016-03-29    0.033280
2016-03-30    0.033324
2016-03-31    0.031662
2016-04-01    0.033901
2016-04-02    0.035763
2016-04-03    0.038823
2016-04-04    0.036628
2016-04-05    0.013170
2016-04-06    0.003171
2016-04-07    0.001352
Name: date_crawled, dtype: float64

From the above information it can be noted that the site was crawled every day for a period of 1 month. Now, calculating the distribution of values in the column ``ad_created`` as percentages

In [121]:
autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
                ...   
2016-04-03    0.039045
2016-04-04    0.037005
2016-04-05    0.011906
2016-04-06    0.003259
2016-04-07    0.001197
Name: ad_created, Length: 76, dtype: float64

From the data above it can be noted that most of the ads were made during the crawling period, and their minority was created a some months earlier. Then calculating the distribution of values in the column ``last_seen`` as percentages

In [122]:
autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index()

2016-03-05    0.001086
2016-03-06    0.004168
2016-03-07    0.005210
2016-03-08    0.007006
2016-03-09    0.009467
2016-03-10    0.010288
2016-03-11    0.012039
2016-03-12    0.023901
2016-03-13    0.008869
2016-03-14    0.012283
2016-03-15    0.015676
2016-03-16    0.016163
2016-03-17    0.027671
2016-03-18    0.007405
2016-03-19    0.015410
2016-03-20    0.020420
2016-03-21    0.020664
2016-03-22    0.021241
2016-03-23    0.018403
2016-03-24    0.019534
2016-03-25    0.018580
2016-03-26    0.016474
2016-03-27    0.015454
2016-03-28    0.020553
2016-03-29    0.021374
2016-03-30    0.024145
2016-03-31    0.023436
2016-04-01    0.022859
2016-04-02    0.024899
2016-04-03    0.024943
2016-04-04    0.024300
2016-04-05    0.126624
2016-04-06    0.225312
2016-04-07    0.134140
Name: last_seen, dtype: float64

From the information on column ``last_seen`` it is possible determine on what day a ad was removed, either because the car was sold or because the ad expired.

### 5.2. Understanding the distribution of 'registration_year'.

In [123]:
autos['registration_year'].describe()

count    45102.000000
mean      2005.062858
std         89.647196
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

From the information above, it is possible to notice the existence of odd values:
- From the information above can be note
- The maximum value is 9999, many years into the future

Once this analysis has been carried out, it is necessary to remove these odd values.

## 6. Dealing with incorrect Registration year data

In first place, to remove the odd values in registration year values, it is necessary perform a counting of number of listings with cars that fall outside some reasonable interval. The range of interval chosen was the 1950 - 2016.

In [124]:
autos["registration_year"].between(1950,2016).value_counts()

True     43311
False     1791
Name: registration_year, dtype: int64

It can be seen that number of cars registrated in the interval of 1900-2016 are  43311 cars while cars outside of that interval sums 1791 cars. Thus the rows of listings with cars that fall outside the interval can be remove from the dataframe.

In [125]:
autos = autos[autos["registration_year"].between(1950,2016)]

In [126]:
autos["registration_year"].value_counts(normalize = True, dropna = False).head(10)

2005    0.066126
2000    0.062686
2004    0.061947
2006    0.061555
2003    0.061347
1999    0.059408
2001    0.058184
2002    0.055852
2007    0.052412
2008    0.050934
Name: registration_year, dtype: float64

It can be seen the most of car are registred in the 2000s

## 7. Exploring Price by Brand

Obtaining the list of brands

In [127]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'renault', 'audi', 'sonstige_autos', 'mazda', 'porsche', 'mini',
       'mercedes_benz', 'seat', 'toyota', 'opel', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'jaguar', 'fiat', 'skoda', 'subaru',
       'kia', 'citroen', 'mitsubishi', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

Now it's interesting to determine the car brands that most appear in the ads

In [128]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211978
bmw               0.115052
mercedes_benz     0.102237
opel              0.100321
audi              0.090531
ford              0.065134
renault           0.043892
peugeot           0.029369
fiat              0.023528
seat              0.017894
skoda             0.017155
smart             0.015123
mazda             0.015008
nissan            0.015008
citroen           0.014153
toyota            0.013415
hyundai           0.010159
sonstige_autos    0.009559
volvo             0.009374
mini              0.009374
honda             0.007804
mitsubishi        0.007735
kia               0.007342
alfa_romeo        0.006626
porsche           0.006419
chevrolet         0.006026
suzuki            0.005911
chrysler          0.003602
dacia             0.002840
jeep              0.002447
land_rover        0.002263
daihatsu          0.002170
subaru            0.001986
saab              0.001662
jaguar            0.001616
daewoo            0.001339
rover             0.001247
t

From the data above can be seen that top 6 most frequent car brands in the ads are **volkswagen**, **bmw**, **mercedes_benz**, **opel**, **audi**, **ford**. Limiting the analysis to these top 6 brands, it is interesting to determine the average price of cars of each brand.

In [132]:
top_brands = autos['brand'].value_counts(normalize=True).index[:6]

In [133]:
mean_price = {}
for brand in top_brands:
    mean_price[brand] = autos.loc[autos["brand"] == brand,"price"].mean()
    
mean_price

{'volkswagen': 5996.269251715499,
 'bmw': 8827.709010636163,
 'mercedes_benz': 8763.703477868112,
 'opel': 3394.7157652474107,
 'audi': 9613.64779393012,
 'ford': 4631.114143920596}

From the information above can be seen that:
- **bmw**, **mercedes_benz** and **audi** are more expensive;
- **opel** and **ford** are less expensive;
- **volkswagen** has an intermediate price.

## 8. Exploring Mileage by Brand

Another valuable piece of information is the average mileage of cars by brand.

In [135]:
mean_mileage = {}
for brand in top_brands:
    mean_mileage[brand] = autos.loc[autos["brand"] == brand,"odometer_km"].mean()
    
mean_mileage

{'volkswagen': 128242.5661692626,
 'bmw': 132894.8424643789,
 'mercedes_benz': 131063.68563685637,
 'opel': 128035.67318757193,
 'audi': 128909.7169089518,
 'ford': 123619.28394186459}

From the information above can be seen that:
- **bmw**, **mercedes_benz** has more average mileage;
- **ford** has the lowest average mileage between the top 6;
- **volkswagen**, **opel**, **audi** has an intermediate average mileage.

# 9. Storing Aggregate Data in a DataFrame

Once the average price and mileage data for the most frequent car brands in the ads has been obtained, it is interesting to group them into a single dataframe to analyze the existence of any relationship between them later.

In [136]:
mean_price_series = pd.Series(mean_price)
mean_mileage_series = pd.Series(mean_mileage)
df_mean_price_mileage = pd.DataFrame(mean_price_series, columns = ['mean_price'])
df_mean_price_mileage["mean_mileage"] =  mean_mileage_series
df_mean_price_mileage

mean_price   mean_mileage
volkswagen     5996.269252  128242.566169
bmw            8827.709011  132894.842464
mercedes_benz  8763.703478  131063.685637
opel           3394.715765  128035.673188
audi           9613.647794  128909.716909
ford           4631.114144  123619.283942

Now it's possible analyzing both mean price and mean mileage for each top car brand in the listing.